onto-viewer-core/src/main/java/org/edmcouncil/spec/ontoviewer/core/utils/StringUtils.java

```diff
@@ -17,7 +17,11 @@ public static String getFragment(IRI iri) {
     if (iriString.contains(AXIOM_PATTERN)) {
       return iriString.substring(iriString.lastIndexOf(".") + 1);
     }
-    return iri.getFragment();
+    String iriFragment = iri.getFragment();
+    if (iriFragment.isEmpty() || iriFragment.isBlank()) {
+      return iriString.substring(iriString.lastIndexOf("/") + 1).replace("#", "");
+    }
+    return iriFragment;
```

Comment:
In my view this string manipulation adds to the technical debt we already have.
In order to control it we need a section in [README.md](https://github.com/edmcouncil/onto-viewer/blob/develop/README.md) that will list the assumptions that we make with respect to IRIs.

Reply:
Take a look at the code and compare it with the previous one, this if is just a solution to a specific case and a problem that occurred, it is not the whole function. The previous solution resulted in an empty text string for ili like `http://www.w3.org/2004/02/skos/core#`. I think the solution used is simple. Then there is no need to make additional assumptions to the appearance of the iri ontology. 

By code, the function now works like this(it is only executed if previous attempts to get the label have failed, the function result is to contain a fragment of iri or a fragment of another identifier):

- iri is reduced to string form for text comparisons(var iriString)
- then checks if it is an axiom (for values like `@viewer.axiom.ClassAssertion`)
    - then it cuts out the appropriate fragment and return
- extract the fragment with owlAPI
- check if it is empty
      -  if it is empty it means that owl api returned an empty string, from the current pr it appears that iri ends in` #` is causing problems, so  it cuts out the appropriate fragment and return
- if it has failed all previous condictions, the iri fragment extracted by owl api above is returned

We have to think about it, but I don't see the downside of solving the problem this way.

Reply:
> Take a look at the code and compare it with the previous one, this if is just a solution to a specific case and a problem that occurred, it is not the whole function. The previous solution resulted in an empty text string for ili like `http://www.w3.org/2004/02/skos/core#`. I think the solution used is simple. Then there is no need to make additional assumptions to the appearance of the iri ontology.
> 
> By code, the function now works like this(it is only executed if previous attempts to get the label have failed, the function result is to contain a fragment of iri or a fragment of another identifier):
> 
> * iri is reduced to string form for text comparisons(var iriString)
> * then checks if it is an axiom (for values like `@viewer.axiom.ClassAssertion`)
>   
>   * then it cuts out the appropriate fragment and return
> * extract the fragment with owlAPI
> * check if it is empty
>   -  if it is empty it means that owl api returned an empty string, from the current pr it appears that iri ends in` #` is causing problems, so  it cuts out the appropriate fragment and return
> * if it has failed all previous condictions, the iri fragment extracted by owl api above is returned
> 
> We have to think about it, but I don't see the downside of solving the problem this way.

The fact that an IRI with '#' separating local fragment creates a problem indicates a problem in the current code, which is heavily dependent on string manipulation. This created a couple of issues in the pasted and we wasted a substantial amount of time to fix them, so it needs to be refactored and not just patched.

Reply:
> Take a look at the code and compare it with the previous one, this if is just a solution to a specific case and a problem that occurred, it is not the whole function. The previous solution resulted in an empty text string for ili like `http://www.w3.org/2004/02/skos/core#`. I think the solution used is simple. Then there is no need to make additional assumptions to the appearance of the iri ontology.
> 
> By code, the function now works like this(it is only executed if previous attempts to get the label have failed, the function result is to contain a fragment of iri or a fragment of another identifier):
> 
> * iri is reduced to string form for text comparisons(var iriString)
> * then checks if it is an axiom (for values like `@viewer.axiom.ClassAssertion`)
>   
>   * then it cuts out the appropriate fragment and return
> * extract the fragment with owlAPI
> * check if it is empty
>   -  if it is empty it means that owl api returned an empty string, from the current pr it appears that iri ends in` #` is causing problems, so  it cuts out the appropriate fragment and return
> * if it has failed all previous condictions, the iri fragment extracted by owl api above is returned
> 
> We have to think about it, but I don't see the downside of solving the problem this way.

Also why not use [getShortForm() ](http://owlcs.github.io/owlapi/apidocs_3/org/semanticweb/owlapi/model/ClassExpressionType.html#getShortForm--) or [getName()](http://owlcs.github.io/owlapi/apidocs_3/org/semanticweb/owlapi/model/ClassExpressionType.html#getName--)? It looks as if we try to reinvent the wheel here.

Reply:
This seemingly works correctly, but in this case the "#" has to be cut out anyway:
![image](https://user-images.githubusercontent.com/26431993/166879268-a22450fd-5014-441b-a2cd-05a62bc1103f.png)

Reply:
> This seemingly works correctly, but in this case the "#" has to be cut out anyway:
> ![image](https://user-images.githubusercontent.com/26431993/166879268-a22450fd-5014-441b-a2cd-05a62bc1103f.png)

Which did you use: getShortForm() or getName()?

Reply:
getShortForm()

Reply:
Try the other one then.

Reply:
> Also why not use [getShortForm() ](http://owlcs.github.io/owlapi/apidocs_3/org/semanticweb/owlapi/model/ClassExpressionType.html#getShortForm--) or [getName()](http://owlcs.github.io/owlapi/apidocs_3/org/semanticweb/owlapi/model/ClassExpressionType.html#getName--)? It looks as if we try to reinvent the wheel here.

getName is an enumeration and specifies the ClassExpressionType, you cannot use an enumeration method in an IRI class object

Reply:
I tested the native function getFragment() of IRI class and it provided the expected results.
```
http://www.w3.org/2004/02/skos/core# > [empty string]
http://www.w3.org/2004/02/skos/core#label > label
http://www.w3.org/2004/02/skos/core/label > label
https://spec.edmcouncil.org/fibo/ontology/FND/Parties/Parties/Undergoer > Undergoer
https://spec.edmcouncil.org/fibo/ontology/FND/Parties/Parties#Undergoer > Undergoer
https://spec.edmcouncil.org/fibo/ontology/FND/Parties/Parties/ > [empty string]
```
@patrycjamia, @dasawanaka Why can't we use it?

Reply:
We need to include this case: 

![image](https://user-images.githubusercontent.com/26431993/168021648-e3d6bdf8-d210-4fca-b4d3-b7f713d63dea.png)

We need to get rid of that # at the end.

Reply:
We need to include this case: 

![image](https://user-images.githubusercontent.com/26431993/168021950-2837aa5a-4dc6-46df-8be8-73913c5b0b6e.png)

We need to get rid of that # at the end.

Reply:
using getFragemnt() returns what is after #.

Reply:
> I tested the native function getFragment() of IRI class and it provided the expected results.
> 
> ```
> http://www.w3.org/2004/02/skos/core# > [empty string]
> ...
> https://spec.edmcouncil.org/fibo/ontology/FND/Parties/Parties/ > [empty string]
> ```
> 
> @patrycjamia, @dasawanaka Why can't we use it?

this is the last function to return a value based on IRI(must return something), if it returns empty text, we are back to the starting point of issue